File: claripy/ast/bits.py
from __future__ import annotations

from typing import cast

from claripy.ast.base import Base
from claripy.errors import ClaripyOperationError


class Bits(Base):
    """
    A base class for AST types that can be stored as a series of bits. Currently, this is bitvectors and IEEE floats.

    :ivar length:       The length of this value in bits.
    """

    __slots__ = ()

    def make_like(self, op, args, **kwargs):
        if "length" not in kwargs:
            kwargs["length"] = self.length
        return Base.make_like(self, op, args, **kwargs)

    def size(self) -> int:
        """
        :returns: The bit length of this AST
        """
        return cast(int, self.length)

    def _type_name(self):
        return self.__class__.__name__ + str(self.length)

    def raw_to_bv(self):
        """
        Converts this data's bit-pattern to a bitvector.
        """
        raise NotImplementedError

    def raw_to_fp(self):
        """
        Converts this data's bit-pattern to an IEEE float.
        """
        raise NotImplementedError

    @staticmethod
    def _check_replaceability(old, new):
        if old.size() != new.size():
            raise ClaripyOperationError("replacements must have matching sizes")

    __len__ = size

File: native/src/ast/bits.cpp
/** @file */
#include "bits.hpp"

#include "../errors/ast.hpp"

#include <sstream>
#include <utility>


// For brevity
using CBits = AST::Cached::Bits;


CBits::Bits(const Ops::Operation o, const Hash h, const Constants::Int l)
    : Base(o, h), length(l) {}

Constants::Int CBits::size() const {
    return this->length;
}

std::string CBits::type_name() const {
    std::stringstream ret;
    ret << this->fundamental_type_name() << this->length;
    return ret.str();
}

std::string CBits::fundamental_type_name() const {
    return "AST::Bits";
}

void CBits::check_replaceability(const AST::Bits &old, const AST::Bits &new_) {
    if (old->size() != new_->size()) {
        throw Errors::factory<Errors::AST::Base>("Replacements must have matching sizes");
    }
}
